### PR TITLE
Introduce TextLocation to track source URI and interval

### PR DESCRIPTION
**Pull request by rafaelbey**
Introduce TextLocation to track source URI and interval

### CODE REVIEW

File: legend-engine-ide-lsp-extension-api/src/main/java/org/finos/legend/engine/ide/lsp/extension/text/TextLocation.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    }
+
+    /**
+     * The source URI for this source
+     * @return source URI
+     */
+    public String getSourceUri()
+    {
+        return sourceUri;
+    }
```
Comment (kevin-m-knight-gs) [inline legend-engine-ide-lsp-extension-api/src/main/java/org/finos/legend/engine/ide/lsp/extension/text/TextLocation.java:42]: Should this be a source URI or just a generic source id? If we're going to call it source URI, should we enforce that it's a valid URI?

File: legend-engine-ide-lsp-extension-api/src/main/java/org/finos/legend/engine/ide/lsp/extension/text/LegendTextObject.java [hunk 2]
```diff
@@ -36,13 +36,13 @@ public abstract class LegendTextObject implements Locatable
      * @param location     object location (required)
      * @param coreLocation "core" location (optional)
      */
-    protected LegendTextObject(TextInterval location, TextInterval coreLocation)
+    protected LegendTextObject(TextLocation location, TextLocation coreLocation)
     {
         this.location = Objects.requireNonNull(location, "location is required");
         this.coreLocation = coreLocation;
-        if ((coreLocation != null) && !location.subsumes(coreLocation))
+        if ((coreLocation != null) && !location.getTextInterval().subsumes(coreLocation.getTextInterval()))
```
[the remark this answers is not in the document]
Comment (rafaelbey) [reply legend-engine-ide-lsp-extension-api/src/main/java/org/finos/legend/engine/ide/lsp/extension/text/LegendTextObject.java:43]: Good catch!  I added subsumes to TextLocation after I changed this class, and forgot to revisit it.  Will change on next commit.

File: legend-engine-ide-lsp-extension-api/src/main/java/org/finos/legend/engine/ide/lsp/extension/text/TextLocation.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+ */
+
+package org.finos.legend.engine.ide.lsp.extension.text;
+
+import java.util.Objects;
+
+/**
+ * A text source location, including the source URI and the text interval on that source URI
+ */
+public class TextLocation implements Comparable<TextLocation>
```
[the remark this answers is not in the document]
Comment (rafaelbey) [reply legend-engine-ide-lsp-extension-api/src/main/java/org/finos/legend/engine/ide/lsp/extension/text/TextLocation.java:24]: Make sense.  Reverting this on my next commit.

File: legend-engine-ide-lsp-extension-api/src/main/java/org/finos/legend/engine/ide/lsp/extension/text/TextLocation.java (hunk 1 above)
Comment (rafaelbey) [reply legend-engine-ide-lsp-extension-api/src/main/java/org/finos/legend/engine/ide/lsp/extension/text/TextLocation.java:42]: We could, but on the eclipse LPS library, source URI's are just represented as String without any further validation.  I'm not sure  about their rational for it, but I think we should be consistent with this.  

I can rename it to source id, or something else if you think make the code more readable.

File: legend-engine-ide-lsp-extension-api/src/main/java/org/finos/legend/engine/ide/lsp/extension/text/TextInterval.java [hunk 4]
```diff
@@ -20,7 +20,7 @@
  * A (continuous) interval of text between two {@link TextPosition}s. Both the start and end positions are inclusive. As
  * such, there is no empty text interval.
  */
-public class TextInterval
+public class TextInterval implements Comparable<TextInterval>
```
[the remark this answers is not in the document]
Comment (rafaelbey) [reply legend-engine-ide-lsp-extension-api/src/main/java/org/finos/legend/engine/ide/lsp/extension/text/TextInterval.java:23]: Make sense.  Reverting this on my next commit.

**Review by kevin-m-knight-gs — APPROVED** [review]

**Pull request merged**